src/scala/microsoft-spark-2.4.x/src/main/scala/org/apache/spark/api/dotnet/CallbackConnection.scala

```diff
@@ -0,0 +1,143 @@
+/*
+ * Licensed to the .NET Foundation under one or more agreements.
+ * The .NET Foundation licenses this file to you under the MIT license.
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.apache.spark.api.dotnet
+
+import java.io.{Closeable, DataInputStream, DataOutputStream}
+import java.net.Socket
+
+import org.apache.spark.internal.Logging
+
+/**
+ * CallbackConnection is used to process the callback communication
+ * between the JVM and Dotnet. It uses a TCP socket to communicate with
+ * the Dotnet CallbackServer and the socket is expected to be reused.
+ * @param address The address of the Dotnet CallbackServer
+ * @param port The port of the Dotnet CallbackServer
+ */
+class CallbackConnection(address: String, port: Int) extends Logging {
+  private[this] val socket = new Socket(address, port)
+  private[this] val inputStream = new DataInputStream(socket.getInputStream)
+  private[this] val outputStream = new DataOutputStream(socket.getOutputStream)
+
+  def send[T](
+    writeBody: DataOutputStream => Unit,
+    readBody: Option[DataInputStream => T]): CallbackResponse[T] = {
+    logInfo("Calling callback...")
+
+    try {
+      SerDe.writeInt(outputStream, CallbackFlags.CALLBACK)
```

Comment:
It will be useful to document the sequence (or layout) of the payload (written/read).

Reply:
but you can do this as a follow up PR (the design doc)